File: kdcreader_ios/ios/Classes/KdcReaderManager.swift
//
//  KdcReaderManager.swift
//  kdcreader_ios
//
//  Created by koamtac on 3/19/24.
//

import Flutter
import Foundation
import CoreBluetooth

class KdcReaderManager {
    // Defer creation and initialization of BLE-related objects
    // Please enable it, if you do not plan to use BLE and do not want to see BLE permission popup.
    private static let USE_LAZY_INIT_FOR_BLE = false
    
    // ** Experimental **
    // Execute commands using disaptch serial queue
    private static let USE_DISPATCH_SERIAL_QUEUE = false

    // Software Decoder using separate notification
    private static let USE_SEPARATE_SW_BARCODE_NOTIFICATION = false
    
    let reader: KDCReader
    
    let converter: KConverter
    var workQueue: DispatchQueue? = nil

    var eventSinkMap = [String: FlutterEventSink]()
    var delegates = [KReaderDelegate]()
    
    // software decoder
    var swDecoderDelegate: KSWDecoderDelegate
    
    init() {
        if KdcReaderManager.USE_LAZY_INIT_FOR_BLE {
//            let options: [String : Any] = [
//                keyLazyInitBLE: true,
//                keyConnectionMode: ConnectionMode.CONNECTION_MODE_ACCESSORY.rawValue
//            ]
            let options: [String : Any] = [
                keyLazyInitBLE: true
            ]
            self.reader = KDCReader(options: options)
        } else {
            self.reader = KDCReader()
        }
        
        self.reader.enableBleAutoReconnect(false)

        self.converter = KConverter()

        if KdcReaderManager.USE_SEPARATE_SW_BARCODE_NOTIFICATION {
            self.reader.enableSoftwareDecoderBarcodeListener(true)
        }

        if KdcReaderManager.USE_DISPATCH_SERIAL_QUEUE {
            self.workQueue = DispatchQueue(label: "koamtac.kdc.sdk.flutter")
        }

        self.swDecoderDelegate = KSWDecoderDelegate(reader: self.reader, converter: self.converter)
        
        let nc = NotificationCenter.default
        
        // Register connection notification
        nc.addObserver(self, selector: #selector(kdcConnectionChanged), name: NSNotification.Name.kdcConnectionChanged, object: nil)
        
        // Register KDC legacy data notification
        nc.addObserver(self, selector: #selector(kdcBarcodeDataArrived), name: NSNotification.Name.kdcBarcodeDataArrived, object: nil)
        nc.addObserver(self, selector: #selector(kdcMSRDataArrived), name: NSNotification.Name.kdcMSRDataArrived, object: nil)
        nc.addObserver(self, selector: #selector(kdcNFCDataArrived), name: NSNotification.Name.kdcNFCDataArrived, object: nil)
        nc.addObserver(self, selector: #selector(kdcDataArrived), name: NSNotification.Name.kdcDataArrived, object: nil)

        // Register KDC mPOS(KDC500) data notification
        nc.addObserver(self, selector: #selector(kposDataArrived), name: NSNotification.Name.kposDataArrived, object: nil)

        // Register error notification
        nc.addObserver(self, selector: #selector(kdcErrorReceived), name: NSNotification.Name.kdcErrorReceived, object: nil)

        // Register information updated notification
        nc.addObserver(self, selector: #selector(kdcInfoUpdated), name: NSNotification.Name.kdcInfoUpdated, object: nil)

        // Register scanned BLE device notification
        nc.addObserver(self, selector: #selector(kdcDeviceScanned), name: NSNotification.Name.kdcDeviceScanned, object: nil)
        
        // Software Decoder When sperate listener used
        nc.addObserver(self, selector: #selector(kdcSwBarcodeDataArrived), name: NSNotification.Name.kdcSwBarcodeDataArrived, object: nil)

        delegates.append(KBLEDelegate(reader: self.reader, converter: self.converter))
        
        delegates.append(KCommonDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KDataProcessDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KSystemDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KBluetoothDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KDisplayDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KVibratorDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KKeypadDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KHidDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KWiFiDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KNfcDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KMsrDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KMsrIcDelegate(reader: self.reader, converter: self.converter))
        delegates.append(KUhfDelegate(reader: self.reader, converter: self.converter))
        
        delegates.append(KPosDelegate(reader: self.reader, converter: self.converter))
        
        delegates.append(swDecoderDelegate)
    }
    
    func execute(_ call: FlutterMethodCall, result: @escaping FlutterResult) -> Bool {
        for delegate in delegates {
            if delegate.isSupported(call.method) {
                if let queue = self.workQueue {
                    queue.async {
                        _ = delegate.run(call, result: result)
                    }
                    return true
                } else {
                    return delegate.run(call, result: result)
                }
            }
        }
        return false
    }
    
    func setEventSink(_ name:String, eventSink: FlutterEventSink?) {
        if let events = eventSink {
            eventSinkMap[name] = events
        } else {
            eventSinkMap.removeValue(forKey: name)
        }
    }
    
    func removeAllEventSink() {
        eventSinkMap.removeAll()
    }

    func setCurrentViewController(viewController: UIViewController) {
        swDecoderDelegate.viewController = viewController
    }

    @objc
    func kdcConnectionChanged(notification: NSNotification) {
        if (notification.object as? KDCReader) != nil {
            if let userInfo = notification.userInfo as? NSDictionary,
                let s = userInfo.object(forKey: keyConnectionState) as? NSNumber,
                let _ = ConnectionState(rawValue: s.intValue) {
                
                var dict = [String: Any]()
                dict[KConstants.ConnectionKey.State] = s
                
                var deviceName: String? = nil

                var device = [String: Any]()
                
                if let accessory = userInfo.object(forKey: keyAccessory) as? EAAccessory {
                    deviceName = String(format: "%@[%@]",
                                        accessory.modelNumber, (accessory.serialNumber as NSString).substring(from: 4))
                    
                    device[KConstants.DeviceKey.DeviceType] = TYPE_ACCESSORY
                    device[KConstants.DeviceKey.SubType] = SUBTYPE_ACCESSORY
                } else if let peripheral = userInfo.object(forKey: keyPeripheral) as? CBPeripheral {
                    deviceName = peripheral.name
                    
                    device[KConstants.DeviceKey.DeviceType] = TYPE_CORE_BLUETOOTH
                    device[KConstants.DeviceKey.SubType] = SUBTYPE_BLUETOOTH_SMART
                    device[KConstants.DeviceKey.Identifier] = peripheral.identifier.uuidString
                } else if let software = userInfo.object(forKey: keySoftware) as? SoftwareDevice {
                    deviceName = software.name
                    device[KConstants.DeviceKey.DeviceType] = TYPE_SOFTWARE
                    device[KConstants.DeviceKey.SubType] = SUBTYPE_SOFTWARE_DECODER
                }
                
                if let name = deviceName {
                    device[KConstants.DeviceKey.DeviceName] = name
                    device[KConstants.DeviceKey.KDCName] = name
                    dict[KConstants.ConnectionKey.Device] = device
                }
                
                if let eventSink = eventSinkMap[KEvent.Name.CONNECTION_STATE_CHANGED] {
                    eventSink(dict)
                }
            }
        }
    }

    @objc
    func kdcBarcodeDataArrived(notification: NSNotification) {
        if let userInfo = notification.userInfo as? NSDictionary,
           let data = userInfo.object(forKey: keyKDCData) as? KDCData {
            let dict = converter.fromKDCData(data)
            
            if let eventSink = eventSinkMap[KEvent.Name.BARCODE_DATA_RECEIVED] {
                eventSink(dict)
            }
        }
    }

    @objc
    func kdcSwBarcodeDataArrived(notification: NSNotification) {
        if let userInfo = notification.userInfo as? NSDictionary,
           let dataList = userInfo.object(forKey: keyKDCSWDataList) as? NSArray {
            var arr = [Dictionary<String, Any>]()
            
            for (index, item) in dataList.enumerated() {
                if let data = item as? KDCData {
                    arr.append(converter.fromKDCData(data))
                }
            }
            
            if let eventSink = eventSinkMap[KEvent.Name.SW_BARCODE_DATA_RECEIVED] {
                eventSink(arr)
            }
        }
    }
    
    @objc
    func kdcMSRDataArrived(notification: NSNotification) {
        if let userInfo = notification.userInfo as? NSDictionary,
           let data = userInfo.object(forKey: keyKDCData) as? KDCData {
            let dict = converter.fromKDCData(data)
            
            if let eventSink = eventSinkMap[KEvent.Name.MSR_DATA_RECEIVED] {
                eventSink(dict)
            }
        }
    }

    @objc
    func kdcNFCDataArrived(notification: NSNotification) {
        if let userInfo = notification.userInfo as? NSDictionary,
           let data = userInfo.object(forKey: keyKDCData) as? KDCData {
            let dict = converter.fromKDCData(data)
            
            if let eventSink = eventSinkMap[KEvent.Name.NFC_DATA_RECEIVED] {
                eventSink(dict)
            }
        }
    }

    @objc
    func kdcDataArrived(notification: NSNotification) {
        if let userInfo = notification.userInfo as? NSDictionary,
           let data = userInfo.object(forKey: keyKDCData) as? KDCData {
            let dict = converter.fromKDCData(data)
            
            if let eventSink = eventSinkMap[KEvent.Name.GENERAL_DATA_RECEIVED] {
                eventSink(dict)
            }
        }
    }

    @objc
    func kposDataArrived(notification: NSNotification) {
        if let data = reader.getPOSData() {
            let dict = converter.fromKPOSData(data)
            
            if let eventSink = eventSinkMap[KEvent.Name.POS_DATA_RECEIVED] {
                eventSink(dict)
            }
        }
    }

    @objc
    func kdcErrorReceived(notification: NSNotification) {
        if let userInfo = notification.userInfo as? NSDictionary,
           let errorCode = userInfo.object(forKey: keyErrorCode) as? Int,
           let errorObject = userInfo.object(forKey: keyErrorObject) as? NSError {
            var dict = [String: Any]()
            dict[KConstants.ResultKey.ErrorCode] = errorCode
            
            let nsErrDict = converter.fromNSError(errorObject)
            dict[KConstants.ResultKey.NSErrorObject] = nsErrDict
            
            if let eventSink = eventSinkMap[KEvent.Name.ERROR_RECEIVED] {
                eventSink(dict)
            }
        }
    }

    @objc
    func kdcInfoUpdated(notification: NSNotification) {
        if let userInfo = notification.userInfo as? NSDictionary,
           let state = userInfo.object(forKey: keyCBManagerState) as? Int {
            var dict = [String: Any]()
            dict[KConstants.InfoUpdatedKey.CbMnagerState] = state
            
            if let eventSink = eventSinkMap[KEvent.Name.INFO_UPDATED] {
                eventSink(dict)
            }
        }
    }
    
    @objc
    func kdcDeviceScanned(notification: NSNotification) {
        if let userInfo = notification.userInfo as? NSDictionary,
           let device = userInfo.object(forKey: keyScannedKdcDevice) as? KDCDevice {
            let dict = converter.fromKDCDevice(device)
            
            if let eventSink = eventSinkMap[KEvent.Name.DEVICE_SCANNED] {
                eventSink(dict)
            }
        }
    }

}

File: kdcreader_android/android/src/main/kotlin/koamtac/kdc/sdk/plugins/KdcReaderManager.kt
package koamtac.kdc.sdk.plugins

import android.app.Activity
import android.content.Context
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbManager
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import koamtac.kdc.sdk.KDCBarcodeDataReceivedListener
import koamtac.kdc.sdk.KDCConnectionListenerEx
import koamtac.kdc.sdk.KDCData
import koamtac.kdc.sdk.KDCDataReceivedListener
import koamtac.kdc.sdk.KDCDevice
import koamtac.kdc.sdk.KDCErrorListener
import koamtac.kdc.sdk.KDCMSRDataReceivedListener
import koamtac.kdc.sdk.KDCNFCDataReceivedListener
import koamtac.kdc.sdk.KDCReader
import koamtac.kdc.sdk.KDCSWBarcodeDataReceivedListener
import koamtac.kdc.sdk.KDCSWData
import koamtac.kdc.sdk.KDCScanResultListener
import koamtac.kdc.sdk.KPOSData
import koamtac.kdc.sdk.KPOSDataReceivedListener
import koamtac.kdc.sdk.plugins.KConstants.ErrorCode
import koamtac.kdc.sdk.plugins.delegate.KBLEDelegate
import koamtac.kdc.sdk.plugins.delegate.KBluetoothDelegate
import koamtac.kdc.sdk.plugins.delegate.KCommonDelegate
import koamtac.kdc.sdk.plugins.delegate.KDataProcessDelegate
import koamtac.kdc.sdk.plugins.delegate.KDisplayDelegate
import koamtac.kdc.sdk.plugins.delegate.KHidDelegate
import koamtac.kdc.sdk.plugins.delegate.KKeypadDelegate
import koamtac.kdc.sdk.plugins.delegate.KMsrDelegate
import koamtac.kdc.sdk.plugins.delegate.KMsrIcDelegate
import koamtac.kdc.sdk.plugins.delegate.KNfcDelegate
import koamtac.kdc.sdk.plugins.delegate.KPosDelegate
import koamtac.kdc.sdk.plugins.delegate.KReaderDelegate
import koamtac.kdc.sdk.plugins.delegate.KSWDecoderDelegate
import koamtac.kdc.sdk.plugins.delegate.KSystemDelegate
import koamtac.kdc.sdk.plugins.delegate.KUhfDelegate
import koamtac.kdc.sdk.plugins.delegate.KVibratorDelegate
import koamtac.kdc.sdk.plugins.delegate.KWiFiDelegate
import koamtac.kdc.sdk.plugins.usb.KUsbDelegate
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch

class KdcReaderManager(val context: Context): KDCConnectionListenerEx,
    KDCBarcodeDataReceivedListener,
    KDCSWBarcodeDataReceivedListener,
    KDCNFCDataReceivedListener,
    KDCMSRDataReceivedListener,
    KDCDataReceivedListener,
    KPOSDataReceivedListener,
    KDCErrorListener {

    companion object {
        // ** Experimental **
        // - Software Decoder using separate notification
        private const val USE_SEPARATE_SW_BARCODE_NOTIFICATION = false

        private const val USB_VENDOR_ID = 7232 // 0x1C40
        private const val USB_PRODUCT_ID = 1081 // 0x0439

        private const val PARAM_NAME_USB_DEVICE = "device"
    }

    private var reader : KDCReader

    // software decoder
    private var swDecoderDelegate : KSWDecoderDelegate

    private var eventSinkMap = HashMap<String, EventSink>()

    private var delegates : ArrayList<KReaderDelegate> = ArrayList()

    init {
        reader = KDCReader().also {
            it.SetContext(context)
            it.EnableBleAutoReconnect(false)

            it.SetKDCConnectionListenerEx(this)

            // KDC Legacy
            it.SetBarcodeDataReceivedListener(this)
            it.SetNFCDataReceivedListener(this)
            it.SetMSRDataReceivedListener(this)
            it.SetDataReceivedListener(this)

            it.SetKDCErrorListener(this)

            // KPOS (KDC500)
            it.SetKPOSDataReceivedListener(this)

            // Software Decoder using separate notification
            if (USE_SEPARATE_SW_BARCODE_NOTIFICATION) {
                it.SetSWBarcodeDataReceivedListener(this)
                it.EnableSoftwareDecoderBarcodeListener(true)
            }
        }

        val bleDelegate = KBLEDelegate(reader)
        bleDelegate.resultListener = object: KDCScanResultListener {
            override fun onDeviceScanned(device: KDCDevice<*>?) {
                val eventMap = KConverter.toMap(device)

                MainScope().launch {
                    val eventSink = eventSinkMap[KEvent.Name.DEVICE_SCANNED]
                    eventSink?.success(eventMap)
                }
            }

            override fun onScanFailed(errCode: Int) {
                val eventMap = HashMap<String, Any>(ErrorCode.instance.FAILED).apply {
                    put(ErrorCode.KEY_OPERATION, errCode)
                }

                MainScope().launch {
                    val eventSink = eventSinkMap[KEvent.Name.DEVICE_SCAN_FAILED]
                    eventSink?.success(eventMap)
                }
            }
        }

        with(delegates) {
            add(bleDelegate)

            add(KCommonDelegate(reader))
            add(KDataProcessDelegate(reader))
            add(KSystemDelegate(reader))
            add(KBluetoothDelegate(reader))
            add(KDisplayDelegate(reader))
            add(KVibratorDelegate(reader))
            add(KKeypadDelegate(reader))
            add(KHidDelegate(reader))
            add(KWiFiDelegate(reader))
            add(KNfcDelegate(reader))
            add(KMsrDelegate(reader))
            add(KMsrIcDelegate(reader))
            add(KUhfDelegate(reader))
            add(KUsbDelegate(reader))

            add(KPosDelegate(reader))

            swDecoderDelegate = KSWDecoderDelegate(reader)
            add(swDecoderDelegate);
        }
    }

    fun execute(call: MethodCall, result: MethodChannel.Result): Boolean {
        for (delegate in delegates) {
            if (delegate.isSupported(call.method)) {
                delegate.run(context, call, result)
                return true
            }
        }
        return false
    }

    fun setEventSink(name:String, eventSink: EventSink?) {
        if(eventSink != null) {
            eventSinkMap[name] = eventSink
        } else {
            eventSinkMap.remove(name)
        }
    }

    fun removeAllEventSink() {
       eventSinkMap.clear()
    }

    // software decoder
    fun setCurrentActivity(activity: Activity?) {
        swDecoderDelegate.activity = activity
    }

    override fun ConnectionChangedEx(device: KDCDevice<*>?, state: Int) {
        val eventMap = HashMap<String, Any>()
        val deviceMap = KConverter.toMap(device)

        eventMap[KConstants.ConnectionKey.Device] = deviceMap
        eventMap[KConstants.ConnectionKey.State] = state

        MainScope().launch {
            val eventSink = eventSinkMap[KEvent.Name.CONNECTION_STATE_CHANGED]
            eventSink?.success(eventMap)
        }
    }

    override fun BarcodeDataReceived(data: KDCData?) {
        val eventMap = KConverter.toMap(data)

        MainScope().launch {
            val eventSink = eventSinkMap[KEvent.Name.BARCODE_DATA_RECEIVED]
            eventSink?.success(eventMap)
        }
    }

    override fun SwBarcodeDataReceived(dataList: MutableList<KDCSWData>?) {
        val eventArr = ArrayList<Map<String, Any?>>()

        if (dataList != null) {
            for (data in dataList) {
               eventArr.add(KConverter.toMap(data))
            }
        }

        MainScope().launch {
            val eventSink = eventSinkMap[KEvent.Name.SW_BARCODE_DATA_RECEIVED]
            eventSink?.success(eventArr)
        }
    }

    override fun NFCDataReceived(data: KDCData?) {
        val eventMap = KConverter.toMap(data)

        MainScope().launch {
            val eventSink = eventSinkMap[KEvent.Name.NFC_DATA_RECEIVED]
            eventSink?.success(eventMap)
        }
    }

    override fun MSRDataReceived(data: KDCData?) {
        val eventMap = KConverter.toMap(data)

        MainScope().launch {
            val eventSink = eventSinkMap[KEvent.Name.MSR_DATA_RECEIVED]
            eventSink?.success(eventMap)
        }
    }

    override fun DataReceived(data: KDCData?) {
        val eventMap = KConverter.toMap(data)

        MainScope().launch {
            val eventSink = eventSinkMap[KEvent.Name.GENERAL_DATA_RECEIVED]
            eventSink?.success(eventMap)
        }
    }

    override fun POSDataReceived(data: KPOSData?) {
        val eventMap = KConverter.toMap(data)

        MainScope().launch {
            val eventSink = eventSinkMap[KEvent.Name.POS_DATA_RECEIVED]
            eventSink?.success(eventMap)
        }
    }

    override fun ErrorReceived(device: KDCDevice<*>?, errCode: Int) {
        val eventMap = HashMap<String, Any>()
        val deviceMap = KConverter.toMap(device)

        eventMap[KConstants.DeviceKey.Device] = deviceMap
        eventMap[KConstants.ResultKey.ErrorCode] = errCode

        MainScope().launch {
            val eventSink = eventSinkMap[KEvent.Name.ERROR_RECEIVED]
            eventSink?.success(eventMap)
        }
    }

    fun handleUsbStateEvent(action: String, device: UsbDevice) {
        if (device.vendorId == USB_VENDOR_ID && device.productId == USB_PRODUCT_ID) {
            val eventMap = KConverter.toMap(device)

            if (action == UsbManager.ACTION_USB_DEVICE_ATTACHED) {
                MainScope().launch {
                    val eventSink = eventSinkMap[KEvent.Name.USB_DEVICE_ATTACHED]
                    eventSink?.success(eventMap)
                }
            }
            if (action == UsbManager.ACTION_USB_DEVICE_DETACHED) {
                if (reader.IsConnected()) {
                    reader.Disconnect()
                }
                MainScope().launch {
                    val eventSink = eventSinkMap[KEvent.Name.USB_DEVICE_DETACHED]
                    eventSink?.success(eventMap)
                }
            }
        }
    }

    fun dispose() {
        eventSinkMap.clear()
        reader.Dispose()
    }
}